lib/tools/node-modules/node-modules-dependencies-builder.ts

```diff
@@ -1,77 +1,63 @@
 import * as path from "path";
-import * as fs from "fs";
+import { NODE_MODULES_FOLDER_NAME, PACKAGE_JSON_FILE_NAME } from "../../constants";
 
 export class NodeModulesDependenciesBuilder implements INodeModulesDependenciesBuilder {
-	private projectPath: string;
-	private rootNodeModulesPath: string;
-	private resolvedDependencies: any[];
-	private seen: any;
-
-	public constructor(private $fs: IFileSystem) {
-		this.seen = {};
-		this.resolvedDependencies = [];
-	}
+	public constructor(private $fs: IFileSystem) { }
 
-	public getProductionDependencies(projectPath: string): any[] {
-		this.projectPath = projectPath;
-		this.rootNodeModulesPath = path.join(this.projectPath, "node_modules");
+	public getProductionDependencies(projectPath: string): IDependencyData[] {
+		const rootNodeModulesPath = path.join(projectPath, NODE_MODULES_FOLDER_NAME);
```

Comment:
Just wanted to share a personal opinion here. So, I think it's been a little bit more complicated than neccessary.

Why can't we combine this in one method and not have nested and hard to debug (for me) calls like:

---- many traverseDependency calls
-------- findModule
------------ readModuleDependencies
---------------- many traverse dependency calls

This nesting sounds too much to me. I'd suggest two approaches for simplification:

### If we want to use recursion, why not rewrite it as: ###

1. many traverseDependency(root_module, depth: 0)
   2.0 if (alreadyResolved(root_module)) return;
   2.1 let currentModule = findCurrentModule
   2.2 resolved[ currentModule ] = currentModule;
   2.3 for(dep in currentModule.dependencies) traverseDependency(dep, depth + 1)

This way we can remove the 3-4 nested methods.

### We can use iteration instead ###
Another suggestion would be to just simulate a queue and turn this into a breadth-first-search, which I would find easier to debug, too. This way we will never care about the `callstack` anymore and we can `debug it` easier

0. for(module in all_initial_modules) resolved[module]  = module
1. queue.push(all_initial_modules)
2. while(queue.length > 0) 
    2.1 let currentModule = queue.shift()
    2.2. for(dep in currentModule.dependencies) 
    2.3.      if (!alreadyResolved(dep)) {
    2.4.           resolved[dep] = dep;
    2.5.           queue.push(dep);
    
Thanks to your tests, we can refactor it at some point if we agree on it.